### PR TITLE
Docker override config from env variables

### PR DESCRIPTION
**Pull request by vk496**
Hello,

It seems a old problem without a clear patch to support it. I've tried to do it as much simple/clean as possible.

With this patch, anyone can override the config with env variable such as:

 - IPFSCONFIG_Swarm_EnableAutoRelay=true
 - IPFSCONFIG_Addresses_Swarm=[\\"/ip4/0.0.0.0/tcp/4001\\",\\"/ip6/::/tcp/4001\\"]

Take in account that:
 - Is case sensitive. Is not the same `IPFSCONFIG_FooBar` than `IPFSCONFIG_FOOBAR`
 - A valid syntax will be accepted even if the path is wrong (`ipfs config` stuff). Something like `IPFSCONFIG_Swarm_EnableAutoRealy=true` will be accepted and written to the config even if it does nothing. So typos could make you loose a lot of hours here :)

Hope it helps you. IPFS is really awesome stuff and now it should be more flexible in container environments

/cc #251 #387 #6325

PS: Docker docs should be updated if is finally merged

### CODE REVIEW

**Comment by aschmahmann** [comment]
Is this necessary given that you can just do `ipfs daemon --init --init-config  <template>`?

**Comment by vk496** [comment]
> Is this necessary given that you can just do `ipfs daemon --init --init-config <template>`?

Most people expect use containers as something fungible. Since you can't apply new config without restarting the daemon you must use volumes.

I think is very very useful be able start a fresh ipfs container with all desired config with just one cammand and ready to use

**Comment by lidel** [comment]
This PR was discussed again during GO triage on 2021-09-03, my quick summary: 

- This will work only for a subset of config options
  - There is a hidden complexity in regard to go-ipfs config file syntax around dynamic maps (DNS, pinning service credentials) which makes us worried about the need for domain-specific syntax here – see https://github.com/ipfs/go-ipfs/pull/8096.  This is not a blocker, but highlights the problem of too many edge cases
- Might create confusion
  - This PR adds env variables that work only with Docker, and not standalone go-ipfs. 
  - Usually, passing env variables acts as an ad-hoc-one-time override, while this PR makes a permanent change to the config. Restarting container without variable does not restore previous value, which is another papercut which we would like to avoid.
  - If we ever add env variable based config override, it should be supported in go-ipfs natively ([`go-ipfs/docs/environment-variables.md`](https://github.com/ipfs/go-ipfs/blob/master/docs/environment-variables.md)), and work as "ad-hoc" override that does not mutate the config file.
  
In short, the decision was to not do this and close this PR without merging, and go with  explicit docker hooks proposed in  https://github.com/ipfs/go-ipfs/pull/8337

Thank you for  the PoC and pushing the discussion forward @vk496 :+1:
